### PR TITLE
add subgroup test for HourAngle

### PR DESCRIPTION
**Pull request by tpolecat**
It occurred to me that we hadn't actually proved `HourAngle` is a subgroup of `Angle` so I added a test for that. Slightly roundabout but the laws might come in handy again if we identify other structure-preserving transformations in the future.

### CODE REVIEW

File: modules/core/shared/src/test/scala/gem/laws/HomomorphismLaws.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+}
+
+abstract class GroupHomomorphismLaws[A, B](f: A => B) extends MonoidHomomorphismLaws[A, B](f) {
+  override val A: Group[A]
+  override val B: Group[B]
+
+  def inverse(a: A): IsEq[B] =
+    f(A.inverse(a)) <-> B.inverse(f(a))
+
```
Comment (sraaphorst) [inline modules/core/shared/src/test/scala/gem/laws/HomomorphismLaws.scala]: This all seems perfect to me. Good stuff.

File: modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gem.laws
+
+// A is a subgroup of B if covariant widening is a group homomorphism.
+abstract class SubgroupLaws[A, B](f: A <:< B) extends GroupHomomorphismLaws[A, B](f)
```
Comment (sraaphorst) [inline modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala]: I assume that you mean by "covariant widening" that you're just talking about the natural map that embeds `A` into `B`. I'm not sure exactly what `<:<` means.

Comment (sraaphorst) [reply modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala]: I'm assuming in this case that you are considering that `A` is a subgroup (not necessarily proper) of `B`?

You're not considering, say, the group `A` of all the multiples of, say, the 2x2 identity matrix under addition  a direct subgroup of the rationals even though you can map `A` isomorphically to the integers, which clearly have an embedding homomorphism into the rationals?

File: modules/core/shared/src/test/scala/gem/laws/discipline/SubgroupTests.scala [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+             ib: Group[B]
+  ): SubgroupTests[A, B] =
+    new SubgroupTests[A, B] {
+      val laws = new SubgroupLaws[A, B](ev) {
+        val A = ia
+        val B = ib
+      }
+    }
+
+}
```
Comment (sraaphorst) [inline modules/core/shared/src/test/scala/gem/laws/discipline/SubgroupTests.scala]: 👍

File: modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala (hunk 2 above)
Comment (tpolecat) [reply modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala]: Yes. In Scala terms

- if `A` is a subtype of `B` (a value of type `A <:< B` is a proof of this fact), and
- we have `Group[A]` and `Group[B]`, and
- we claim that `A` is a subgroup of `B`, then

the widening `A => B` (i.e., just forgetting it's an `A`) must be a group homomorphism.

**Comment by tpolecat** [comment]
There was a problem with the `Index` spec, which I fixed here. Will need re-approval when it's green.

File: modules/core/shared/src/test/scala/gem/math/HourAngleSpec.scala [hunk 4]
```diff
@@ -17,6 +17,7 @@ final class HourAngleSpec extends CatsSuite {
   // Laws
   checkAll("HourAngle", CommutativeGroupTests[HourAngle].commutativeGroup)
   checkAll("HourAngle", EqTests[HourAngle].eqv)
+  checkAll("HourAngle <: Angle", SubgroupTests[HourAngle, Angle].subgroup)
```
Comment (sraaphorst) [inline modules/core/shared/src/test/scala/gem/math/HourAngleSpec.scala]: I needed to read your description about what exactly an `HourAngle` was to see why it wasn't isomorphic to `Angle`, but now I get it. Yes, it should definitely embed via the identity homomorphism as a subgroup of `Angle`.

File: modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala (hunk 2 above)
Comment (tpolecat) [reply modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala]: > You're not considering ...

Yes, if I understand what you're saying. Here every `A` *is* a `B`; it's a subset. So this isn't an arbitrary mapping.

Comment (sraaphorst) [reply modules/core/shared/src/test/scala/gem/laws/SubgroupLaws.scala]: That sounds exactly right to me. Thanks for the explanation regarding `<:<`.

File: modules/core/shared/src/test/scala/gem/math/HourAngleSpec.scala (hunk 4 above)
Comment (tpolecat) [reply modules/core/shared/src/test/scala/gem/math/HourAngleSpec.scala]: 👍

**Pull request merged**